crates/sui-framework/sources/LockedCoin.move

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module Sui::LockedCoin {
+    use Sui::Balance::{Self, Balance};
+    use Sui::Coin::{Self, Coin};
+    use Std::Errors;
+    use Sui::ID::{Self, VersionedID};
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+
+    /// The locked_until time passed into the creation of a locked coin is invalid.
+    const EINVALID_LOCK_UNTIL: u64 = 0;
+    /// Attempt is made to unlock a locked coin that has not been unlocked yet.
+    const ECOIN_STILL_LOCKED: u64 = 1;
+
+    /// A coin of type `T` locked until `locked_until_epoch`.
+    struct LockedCoin<phantom T> has key, store {
+        id: VersionedID,
+        balance: Balance<T>,
+        locked_until_epoch: u64
+    }
+
+    /// Returns the epoch until which the coin is locked.
+    public fun locked_until_epoch<T>(locked_coin: &LockedCoin<T>) : u64 {
+        locked_coin.locked_until_epoch
+    }
+
+    /// Wrap a balance into a LockedCoin.
+    public fun from_balance<T>(balance: Balance<T>, locked_until_epoch: u64, ctx: &mut TxContext): LockedCoin<T> {
```

Comment:
We might want a constructor that asserts `locked_until_epoch > ctx.epoch` as a sanity check.

Reply:
Actually we may not want this. Right now this `from_balance` function is used by the undelegation logic to give back the delegated stake. Suppose someone delegates their locked SUI to a validator and later calls `undelegate` to get their locked coin back after `locked_until_epoch` has passed.  Do we want the undelegation logic to automatically unlock the locked coin? It's probably simpler to keep unlocking and undelegation separate.